### ISSUE TITLE
Accumulated gradients

### ISSUE THREAD

**Issue by ebilal** [facebookresearch/XLM#217]
Is training with --batch_size 64 equivalent to training with --batch_size 8 and --accumulate_gradients 8 ? I find that the second one converges slower. I did account for the 8x increase in the number of iterations when doing the comparison. Do I need to change the optimizer parameters? (lr, warm-up) Thanks!

**Comment by glample** [comment]
If you use Adam, it should be identical. `--accumulate_gradients 8` will result in gradients 8x larger, but this fine since Adam is invariant by gradient constant rescaling. With SGD you would need to update the learning rate (8x slower with accumulation).
How did you take into account the number of iterations when doing the comparison? Be careful, because one epoch is a number of sentences per GPU. So if you double the number of GPUs, you see 2x more data at the end of an epoch (it is a slightly unusual way to measure epochs, but much more convenient as we don't have to adjust the epoch size when the number of GPU changes).

**Comment by ebilal** [comment]
So running on two GPUs with batch_size 32 for one epoch is equivalent to running on one GPU with batch_size 32 and accumulate_gradients 2 for two epochs?

**Comment by glample** [comment]
Basically you have:
number of updates = n_epochs * (epoch_size / batch_size) / accumulate_gradients
effective batch_size = n_gpu * accumulate_gradients

In the first case you have:
number of updates = 1 * (epoch_size / batch_size) / 1
effective batch_size = 2 * 1

In the second case you have:
number of updates = 2 * (epoch_size / batch_size) / 2
effective batch_size = 1 * 2

So yes, this should be equivalent.

**Comment by formiel** [comment]
> effective batch_size = n_gpu * accumulate_gradients

@glample  Shouldn't it be: `effective_batch_size = n_gpu * accumulate_gradients * batch_size`?

**Comment by glample** [comment]
Yes you are correct.
